roboRIO-2024/src/main/java/frc/robot/subsystems/OdometrySubsystem.java

```diff
@@ -0,0 +1,132 @@
+import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
+import edu.wpi.first.units.BaseUnits;
+import edu.wpi.first.units.Unit;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.smartdashboard.Field2d;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.RobotContainer;
+import frc.robot.Settings;
+
+/**<h1>i hate this **/
+public class OdometrySubsystem extends SubsystemBase{
+
+
+  Pose2d pose = new Pose2d();
+
+  PigeonIMU gyro;
+  RelativeEncoder leftEncoder;
+  RelativeEncoder rightEncoder;
+  Drivetrain drivetrain;
+    
+  Field2d field2d = new Field2d();
+
+  PIDController leftController = new PIDController(0.2, 0, 0.02);
+  PIDController rightController = new PIDController(0.2, 0, 0.02);
+
+  private static final double kTrackWidth = 0.381 * 2; // meters, this is the defauklt from wpilib
+                                                       // change this later
+
+  private final DifferentialDriveKinematics kinematics =
+    new DifferentialDriveKinematics(kTrackWidth);
+
+    
+  private final SimpleMotorFeedforward feedforward = new SimpleMotorFeedforward(1, 3);
+    
+    // constructor so i can find in in the wall of code
+  public OdometrySubsystem (Drivetrain drivetrain) {
+      this.drivetrain = drivetrain;
+      gyro = drivetrain.getGyro();
+      leftEncoder = drivetrain.getLeftEncoder();
+      rightEncoder = drivetrain.getRightEncoder();
+
+      AutoBuilder.configureRamsete(
+        this::getPose, // Robot pose supplier
+        this::resetPose, // Method to reset odometry (will be called if your auto has a starting pose)
+        this::getCurrentSpeeds, // Current ChassisSpeeds supplier
+        this::driveChassisSpeeds, // Method that will drive the robot given ChassisSpeeds
+        new ReplanningConfig(), // Default path replanning config. See the API for the options here
+        () -> {
+                    // Boolean supplier that controls when the path will be mirrored for the red alliance
+                    // This will flip the path being followed to the red side of the field.
+                    // THE ORIGIN WILL REMAIN ON THE BLUE SIDE
+
+          var alliance = DriverStation.getAlliance();
+          if (alliance.isPresent()) {
+            return alliance.get() == DriverStation.Alliance.Red;
+          }
+          return false;
+        },
+        this // Reference to this subsystem to set requirements
+        );
+    }
+
+  DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(
+    new Rotation2d(getAngle()),
+    leftEncoder.getPosition(), rightEncoder.getPosition(),
+    new Pose2d(5.0, 13.5, new Rotation2d()));
+
+  @Override
+  public void periodic() {
+    pose = m_odometry.update(new Rotation2d(getAngle()),
+    leftEncoder.getPosition(),
```

Comment:
You should also have to setPositionConversionFactor. Also, you should extract all position and velocity calls to a function.